middle_end/flambda2/from_lambda/closure_conversion_aux.ml

```diff
@@ -719,39 +723,53 @@ end
 
 module Let_with_acc = struct
   let create acc let_bound named ~body =
-    let cost_metrics_of_defining_expr =
-      match (named : Named.t) with
-      | Prim (prim, _) -> Code_size.prim prim |> Cost_metrics.from_size
-      | Simple simple -> Code_size.simple simple |> Cost_metrics.from_size
-      | Static_consts _consts -> Cost_metrics.zero
-      | Set_of_closures set_of_closures ->
-        let code_mapping = Acc.code acc in
-        Cost_metrics.set_of_closures
-          ~find_code_characteristics:(fun code_id ->
-            let code = Code_id.Map.find code_id code_mapping in
-            { cost_metrics = Code.cost_metrics code;
-              params_arity = List.length (Code.params_arity code)
-            })
-          set_of_closures
-      | Rec_info _ -> Cost_metrics.zero
+    let is_unused_singleton =
+      match Bound_pattern.must_be_singleton_opt let_bound with
+      | Some var ->
+        not (Name_occurrences.mem_var (Acc.free_names acc) (Bound_var.var var))
+      | None -> false
     in
-    let acc =
-      Acc.increment_metrics
-        (Cost_metrics.increase_due_to_let_expr ~is_phantom:false
-           ~cost_metrics_of_defining_expr)
-        acc
-    in
-    let free_names_of_body = Or_unknown.Known (Acc.free_names acc) in
-    let acc =
-      Bound_pattern.fold_all_bound_names let_bound ~init:acc
-        ~var:(fun acc var ->
-          Acc.remove_var_from_free_names (Bound_var.var var) acc)
-        ~symbol:(fun acc s -> Acc.remove_symbol_from_free_names s acc)
-        ~code_id:(fun acc cid -> Acc.remove_code_id_from_free_names cid acc)
+    let has_no_effects =
+      match (named : Named.t) with
+      | Prim (prim, _) -> Flambda_primitive.no_effects_or_coeffects prim
```

Comment:
It's fine to remove expressions with only coeffects. In fact, it should be fine to remove expressions with only generative effects too.